File: sem4/jpp/list4/cpp/main.cpp
#include "Philosopher.hpp"

#include <mutex>
#include <thread>
#include <vector>

constexpr int NUMBER_OF_PHILOSOPHERS = 5;
constexpr int MEAL_LIMIT = 3;

int main() {
    std::vector<std::mutex> chopsticks(NUMBER_OF_PHILOSOPHERS);
    std::vector<std::thread> philosophers(NUMBER_OF_PHILOSOPHERS);
    std::mutex printingMutex;

    for (int i = 0; i < NUMBER_OF_PHILOSOPHERS; i++) {
        auto& leftChopstick = chopsticks[i];
        auto& rightChopstick = chopsticks[(i + 1) % NUMBER_OF_PHILOSOPHERS];
        Philosopher philosopher = (i % 2 == 0)
                    ? Philosopher(i, rightChopstick, leftChopstick, MEAL_LIMIT, printingMutex)
                    : Philosopher(i, leftChopstick, rightChopstick, MEAL_LIMIT, printingMutex);
        philosophers[i] = std::thread(philosopher);
    }

    for (auto &philosopherThread : philosophers) {
        philosopherThread.join();
    }

    return 0;
}

File: sem4/jpp/list4/go/main.go
package main

import (
	"dining-philosophers/philosopher"
	"sync"
)

const numberOfPhilosophers int = 5
const mealLimit int = 3

var waitGroup = sync.WaitGroup{}

func main() {
	var chopsticks = [numberOfPhilosophers]sync.Mutex{}
	var philosophers = [numberOfPhilosophers]*philosopher.Philosopher{}

	for i := 0; i < numberOfPhilosophers; i++ {
		var leftChopstick = &chopsticks[i]
		var rightChopstick = &chopsticks[(i+1)%numberOfPhilosophers]
		var phil *philosopher.Philosopher
		if i%2 == 0 {
			phil = philosopher.NewPhilosopher(i, rightChopstick, leftChopstick, mealLimit, &waitGroup)
		} else {
			phil = philosopher.NewPhilosopher(i, leftChopstick, rightChopstick, mealLimit, &waitGroup)
		}
		philosophers[i] = phil
	}

	for _, phil := range philosophers {
		waitGroup.Add(1)
		go phil.Dine()
	}
	waitGroup.Wait()
}
